src/execution_plan/ops/op.h

```diff
@@ -70,30 +68,35 @@ typedef struct {
 	double profileExecTime;     // Operation total execution time in ms.
 }  OpStats;
 
+struct ExecutionPlan;
+
 struct OpBase {
-	OPType type;                // Type of operation
+	OPType type;                // Type of operation.
 	fpInit init;                // Called once before execution.
+	fpFree free;                // Free operation.
+	fpReset reset;              // Reset operation state.
 	fpConsume consume;          // Produce next record.
 	fpConsume profile;          // Profiled version of consume.
-	fpReset reset;              // Reset operation state.
-	fpFree free;                // Free operation.
 	fpToString toString;        // operation string representation.
 	char *name;                 // Operation name.
-	uint *modifies;             // List of Record indices this op modifies.
-	RecordMap *record_map;      // Mapping of entities into Record IDs in the scope of this ExecutionPlanSegment.
-	struct OpBase **children;   // Child operations.
 	int childCount;             // Number of children.
 	bool op_initialized;        // True if the operation has already been initialized.
+	struct OpBase **children;   // Child operations.
+	const char **modifies;      // List of entities this op modifies.
 	OpStats *stats;             // Profiling statistics.
 	Record *dangling_records;   // Records allocated by this operation that must be freed.
 	struct OpBase *parent;      // Parent operations.
+	const struct ExecutionPlan *plan; // ExecutionPlan this operation is part of.
 };
 typedef struct OpBase OpBase;
 
-void OpBase_Init(OpBase *op);       // Initialize op.
+// Initialize op.
+void OpBase_Init(OpBase *op, OPType type, char *name, fpInit init, fpConsume consume, fpReset reset,
+				 fpToString toString, fpFree free, const struct ExecutionPlan *plan);
 void OpBase_Free(OpBase *op);       // Free op.
 Record OpBase_Consume(OpBase *op);  // Consume op.
 Record OpBase_Profile(OpBase *op);  // Profile op.
+
 int OpBase_ToString(const OpBase *op, char *buff, uint buff_len);
 
 /* If an operation holds the sole reference to a Record it is evaluating,
```

Comment:
Yes, very much - every operation that calls AR_EXP_Evaluate registers its Records this way so that they can be freed properly in the event of run-time errors. It's...not an ideal solution.